File: Algorithm/HeapSort.m
% ∂—≈≈–Ú£¨‘≠÷∑
function Array=HeapSort(Array)
len=length(Array);
heapsize=len;
Array=sub_buildmaxHeap(Array);

for i=len:-1:2
    temp=Array(i);
    Array(i)=Array(1);
    Array(1)=temp;
    heapsize=heapsize-1;
    Array=sub_maxHeapify(Array,1,heapsize);
end
end

function A=sub_maxHeapify(A,i,heapsize)
Left=@(i) 2*i;
Right=@(i) 2*i+1;

l=Left(i);
r=Right(i);

if l<=heapsize && A(l)>A(i)
    largest=l;
else
    largest=i;
end
if r<=heapsize && A(r)>A(largest)
    largest=r;
end
if largest~=i
    temp=A(i);
    A(i)=A(largest);
    A(largest)=temp;
    A=sub_maxHeapify(A,largest,heapsize);
end
end

function A=sub_buildmaxHeap(A)
heapsize=length(A);
for i=floor(heapsize/2):-1:1
    A=sub_maxHeapify(A,i,heapsize);
end
end

File: Algorithm/Sort & Order Statistic/C/HeapSort.c
void sub_maxHeapify(int* A, int i, int heapsize)
{
	int l = 2 * (i + 1) - 1;
	int r = 2 * (i + 1);
	int largest, temp;
	if (l < heapsize && *(A + l) > *(A + i))
		largest = l;
	else
		largest = i;
	if (r < heapsize && *(A + r) > *(A + largest))
		largest = r;
	if (largest != i)
	{
		temp = *(A + i);
		*(A + i) = *(A + largest);
		*(A + largest) = temp;
		sub_maxHeapify(A, largest, heapsize);
	}
}

void sub_buildmaxHeap(int* A, int heapsize)
{
	for (int i = (heapsize - 1) / 2; i >= 0; i--)
		sub_maxHeapify(A, i, heapsize);
}

void HeapSort(int* A, int arraysize)
{
	int temp;
	int heapsize = arraysize;
	sub_buildmaxHeap(A, heapsize);
	for (int i = arraysize - 1; i > 0; i--)
	{
		temp = *(A + i);
		*(A + i) = *A;
		*A = temp;
		heapsize--;
		sub_maxHeapify(A, 0, heapsize);
	}
}